source/vault/helpers.go

```diff
@@ -1,57 +1,98 @@
 package vault
 
 import (
+	"encoding/json"
+	"errors"
+	"io/ioutil"
 	"os"
 
 	"github.com/MiLk/nsscache-go/source"
 	"github.com/hashicorp/vault/api"
 )
 
-func CreateVaultSource(prefix string) (source.Source, error) {
-	client, err := CreateVaultClient()
+type wrappedData struct {
+	Token           string `json:"token"`
+	Accessor        string `json:"accessor"`
+	TTL             int    `json:"ttl"`
+	CreationTime    string `json:"creation_time"`
+	CreationPath    string `json:"creation_path"`
+	WrappedAccessor string `json:"wrapped_accessor"`
+}
+
+// CreateVaultSource returns a vault source with a client associated to work with
+func CreateVaultSource(prefix string, fpath string) (source.Source, error) {
+	client, err := CreateVaultClient(fpath)
 	if err != nil {
 		return nil, err
 	}
 	return NewSource(Client(client), Prefix(prefix))
 }
 
-func CreateVaultClient() (*api.Client, error) {
+// CreateVaultClient returns a Vault Client with a valid Token provided by the Vault Agent assigned to it.
+//
+// `fpath` indicates the path of the file to read from. This file is where the token provided by the agent is stored.
+func CreateVaultClient(fpath string) (*api.Client, error) {
 	client, err := api.NewClient(nil)
 	if err != nil {
 		return nil, err
 	}
-	// If there is no token read from the environment
-	if client.Token() == "" {
-		ghToken := os.Getenv("VAULT_AUTH_GITHUB_TOKEN")
-		if ghToken != "" {
-			// Try to authenticate with GitHub if there is a token
-			secret, err := client.Logical().Write("auth/github/login", map[string]interface{}{
-				"token": ghToken,
-			})
-			if err != nil {
-				return nil, err
-			}
-			client.SetToken(secret.Auth.ClientToken)
-		} else {
-			// Try to authenticate with EC2
-			pkcs7, err := getPkcs7()
-			if err != nil {
-				return nil, err
-			}
-			nonce, err := getNonce("/etc/vault-nonce")
-			if err != nil {
-				return nil, err
-			}
-			secret, err := client.Logical().Write("auth/aws-ec2/login", map[string]interface{}{
-				"role":  "nsscache",
-				"pkcs7": pkcs7,
-				"nonce": nonce,
-			})
-			if err != nil {
-				return nil, err
-			}
-			client.SetToken(secret.Auth.ClientToken)
+
+	rawToken, err := ReadToken(fpath)
+	if err != nil {
+		return nil, err
+	}
+
+	var wrappedData wrappedData
+	var token string
+
+	// Check if the token has been stored in JSON format (wrapped token) or as a plain string
+	if err := json.Unmarshal(rawToken, &wrappedData); err == nil {
```

Comment:
Can't you just check the first byte of `rawToken` to detect if it's wrapped or not?
And then only if it's wrapped, you unmarshal.
Right now, if it's wrapped, but for some reason the unmarshaling fails, you will end up with a wring token.